### PR TITLE
JAMES-3945 Subaddressing - case sensitivity and subfolders

### PR DESCRIPTION
**Pull request by florentos17**
1) Currently we have a case sensitivity problem: folders are case sensitive whereas email addresses are not. We would want both `fazavant+folder@domain` and `fazavant+Folder@domain` to be delivered in my folder `Folder`. Moreover, if both `folder` and `Folder` exist, the correct one should be selected.

2) Since we agreed to give the `p` right to either noone or anyone, there is no need for unit tests about whitelists and blacklists: I removed them

3) We want to check whether a subfolder can be subaddressed. I wrote a unit test that sends an email to `fazavant+folder/subfoder@domain`, and it seems to be all good :) is that enough or should I write more advanced tests ?

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java [hunk 1]
```diff
@@ -100,10 +101,62 @@ void subAddressedEmailShouldBeDeliveredInINBOXWhenSpecifiedFolderDoesNotExist(@T
...
+
+    @Test
+    void subAddressedEmailShouldBeDeliveredInSpecifiedFolderWithCorrectUpperCaseWhenSeveralCasesExist(@TempDir File temporaryFolder) throws Exception {
+        setup(temporaryFolder);
+
+        // create mailboxes
+        testIMAPClient.sendCommand("CREATE " + TARGETED_MAILBOX);
+        testIMAPClient.sendCommand("SETACL " + TARGETED_MAILBOX + " " + "anyone" + " p");
+        testIMAPClient.sendCommand("CREATE " + TARGETED_MAILBOX_UPPER);
+        testIMAPClient.sendCommand("SETACL " + TARGETED_MAILBOX_UPPER + " " + "anyone" + " p");
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java:154]:
```suggestion
        testIMAPClient.sendCommand("CREATE " + TARGETED_MAILBOX);
        testIMAPClient.sendCommand("SETACL " + TARGETED_MAILBOX + " " + "anyone" + " p");
        testIMAPClient.sendCommand("CREATE " + TARGETED_MAILBOX_UPPER);
```

Then what happens?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SubAddressing.java [hunk 2]
```diff
@@ -83,33 +86,51 @@ public SubAddressing(UsersRepository usersRepository, @Named("mailboxmanager") M
...
     }
 
-    private void postIfHasRight(Mail mail, MailAddress recipient, String targetFolder) throws UsersRepositoryException, MailboxException {
-        if (hasPostRight(mail, recipient, targetFolder)) {
-            StorageDirective.builder().targetFolders(ImmutableList.of(targetFolder)).build()
+    private Optional<MailboxPath> getPathWithCorrectCase(MailAddress recipient, String targetFolder) throws UsersRepositoryException, MailboxException {
+        Username recipientUsername = usersRepository.getUsername(recipient);
+        MailboxSession session = mailboxManager.createSystemSession(recipientUsername);
+
+        return mailboxManager.list(session).stream()
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SubAddressing.java:99]: search(MailboxQuery) !!!!!!

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java (hunk 1 above)
Comment (florentos17) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java:154]: the desired outcome was already correctly achieved, but i added a test for that specific case to make it clear :)

**Review by Arsnael — COMMENTED** [review]
Read it nothing to add

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/search/ExactNameCaseIncensitive.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.model.search;
+
+import com.google.common.base.Preconditions;
+
+import java.util.Objects;
+
+public class ExactNameCaseIncensitive implements MailboxNameExpression {
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/search/ExactNameCaseIncensitive.java:26]: I do not see where case insentivity is?

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/search/ExactNameCaseIncensitive.java:26]: Unit tests for these class?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/search/PrefixedWildcardCaseIncensitive.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.model.search;
+
+import com.google.common.base.Preconditions;
+
+import java.util.Objects;
+
+public class PrefixedWildcardCaseIncensitive implements MailboxNameExpression {
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/search/PrefixedWildcardCaseIncensitive.java:26]: Where is this used?

**Comment by florentos17** [comment]
i created a new implementation of `MailboxNameExpression` to find only the mailboxes that match the name without taking the case into account - it is almost identical to the existing `ExactName` except for the `isExpressionMatch` and `includeChildren` methods.
i added unit tests which are also extremely similar to those of `ExactName`: just a couple of extra tests to test various cases.

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**